### PR TITLE
tarfile improvements

### PR DESCRIPTION
**Pull request by srittau**
* Use `Literal` for `mode`.
* Don't explicitly derive from `Iterable`.
* Use imports from `collections.abc`.
* Partialy replace `IO` in `TarFile`.

More `IO` improvements are possible in the future:

* Split into `_ReadableFileobj` and `_WritableFileobj`, depending on `mode`.
* `TarFile.bz2open()` and `.xzopen()` still use `IO`. This depends on improving `bz2.pyi` and `lzma.pyi` respectively gaining protocols.
* `TarFile.open()` and module-level `open()` still use `IO`. This could be improved using overloads.
* Some other functions and methods still use `IO`.

Closes: #5491

### CODE REVIEW

**Comment by srittau** [comment]
GitHub Actions is having issues again. https://www.githubstatus.com/

**Comment by srittau** [comment]
It passes now.

**Comment by JelleZijlstra** [comment]
I need more time to look at this properly, but it feels like a shame if we don't switch `tarfile.open` to a protocol, since that's presumably the main API users would use; it's also the API being used in the linked issue.

**Comment by srittau** [comment]
Good point about this not closing the linked issue. `tarfile.open()` (which is alias of `tarfile.TarFile.open()`) is certainly doable and should be done, but needs may overloads and protocols for `lzma` and `bz2` to be fully typed.

**Comment by srittau** [comment]
To clarify: If I find time to do more, I'd like to do it in a separate PR. And possibly work on `lzma` and `bz2` first.

File: stdlib/tarfile.pyi [hunk 1]
```diff
@@ -70,11 +83,11 @@ def open(
 class ExFileObject(io.BufferedReader):
     def __init__(self, tarfile: TarFile, tarinfo: TarInfo) -> None: ...
 
-class TarFile(Iterable[TarInfo]):
+class TarFile:
     OPEN_METH: Mapping[str, str]
     name: Optional[StrOrBytesPath]
-    mode: str
-    fileobj: Optional[IO[bytes]]
+    mode: Literal["r", "a", "w"]
```
Comment (JelleZijlstra) [inline stdlib/tarfile.pyi:89]: Looks like the implementation also allows `"x"`.
```suggestion
    mode: Literal["r", "a", "w", "x"]
```

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Looks good apart from one detail

**Pull request merged**